src/core/batch/batch.ts

```diff
@@ -20,47 +23,115 @@ interface BloomFilterData {
+    this.name = "MixedTypeBatchError";
+  }
+}
+
+type BatchIterable<T extends DSNPType> = AsyncOrSyncIterable<DSNPMessageSigned<DSNPTypedMessage<T>>>;
+
 /**
- * createFile() takes a series of DSNP messages and returns a URL
- * for storage location.
+ * createFile() takes a series of Batch DSNP messages, writes them to a file at
+ * specified target path and returns a BatchFileData object.
  *
  * @param targetPath - The path to and name of file
+ * @param dsnpType - The DSNPType of the messages provided
  * @param messages - An array of DSNPMessage to include in the batch file
  * @param opts - Optional. Configuration overrides, such as store, if any
- * @returns A URL of the storage location
- * @throws error if messages argument is empty.
+ * @returns A BatchFileData object including a URL and keccak hash of the file
  */
-export const createFile = async (targetPath: string, messages: DSNPMessage[], opts?: ConfigOpts): Promise<URL> => {
-  if (messages.length === 0) throw EmptyArrayError;
-
-  const schema = new ParquetSchema(getSchemaFor(messages[0].dsnpType));
-  const bloomFilterOptions = getBloomFilterOptionsFor(messages[0].dsnpType);
+export const createFile = async <T extends DSNPType>(
+  targetPath: string,
+  dsnpType: T,
+  messages: BatchIterable<T>,
+  opts?: ConfigOpts
+): Promise<BatchFileData> => {
+  const schema = new ParquetSchema(getSchemaFor(dsnpType));
+  const bloomFilterOptions = getBloomFilterOptionsFor(dsnpType);
 
   const store = requireGetStore(opts);
-  return store.putStream(targetPath, async (writeStream: WriteStream) => {
-    await writeBatch(writeStream, schema, messages, bloomFilterOptions);
+  const hashGenerator = keccak256.create();
+  const url = await store.putStream(targetPath, async (writeStream: WriteStream) => {
+    const hashingWriteStream = {
+      ...writeStream,
+      write: (chunk: Uint8Array, ...args: unknown[]): boolean => {
+        hashGenerator.update(chunk);
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        return writeStream.write(chunk, ...(args as any[]));
+      },
+    };
+    await writeBatch(hashingWriteStream, schema, messages, bloomFilterOptions);
   });
+
+  return {
+    url,
+    hash: hashGenerator.hex(),
+  };
 };
 
 /**
- * writeBatch() takes a series of DSNP messages and returns a Batch file
- * object for publishing.
+ * writeBatch() takes a series of Batch DSNP messages, writes them to the given
+ * stream and returns a void promise which resolves when done.
  *
  * @param writeStream - A writable stream
  * @param schema - The ParquetJS schema for the messages DSNP type
  * @param messages - An array of DSNPMessage to include in the batch file
  * @param opts - Options for creating a Parquet file
  * @returns A void promise which will either resolve or reject
  */
-export const writeBatch = async (
+export const writeBatch = async <T extends DSNPType>(
```

Comment:
DSNPType is an enum can you please explain what this does? I am still a newbie at Typescript.
Why not just use DSNPType?

Reply:
This generic allows us to get specific about which type of BatchIterable we have. Although in this case we don't need it because only part of each message we use is `msg.dsnpType`. So we could do this instead I think:

```suggestion
export const writeBatch = async (
  writeStream: WriteStream,
  schema: Schema,
  messages: BatchIterable<DSNPType>,
```

Reply:
My intention with this generic was to add a type check that messages iterable passed to this function only ever has one DSNPType, i.e. that it's something like `[Broadcast, Broadcast, Broadcast]` not `[Broadcast, Reply, Broadcast]`.

I know this check only applies if the user is also using TypeScript, so I also added a logical check in the body of the function below which throws `MixedTypeBatchError` if multiple are passed, but my thinking was that *if* the SDK user is also using TypeScript, they can avoid that error entirely with some extra type safety.

Do we want to drop this generic and lose that type safety?

Reply:
That is a really good point. I had forgotten about that type check.

Keep the safety!